File: 2011-mebipenny/contest/hexcode/hexcode.hs
count_interpretations xs 0 0 = 0
count_interpretations [] a b = a
count_interpretations (x:[]) a b = if (isvalid_one_char x) then (a + b) else b
count_interpretations (x:y:xs) a b = count_interpretations (y:xs) new_a new_b
    where
        new_b = if (isvalid_two_char x y) then a else 0
        new_a = if (isvalid_one_char x) then (a + b) else b

isvalid_two_char x y = (num >= 1) && (num <= 26) && (x /= '0')
    where
        num = (read ("0x" ++ x:y:[])::Int)

isvalid_one_char x = (num >= 1) && (num <= 26)
    where
        num = (read ("0x" ++ x:[])::Int)

handleInput "0" = do return ()
handleInput x = do
    print $ count_interpretations x 1 0
    main

main = do
    line <- getLine
    handleInput line

File: 2011-mebipenny/contest/hexcode/hexcode.rb

def is_valid_one_char(char)
  char = char.hex
  return char >= 1 && char <= 26
end

def is_valid_two_char(chars)
  char = chars.hex
  return char >= 1 && char <= 26 && chars[0..0] != "0"
end

def count_interpretations(xs, a, b)
  return 0 if a == 0 and b == 0
  return a if xs.size == 0
  if xs.size == 1
    return a + b if is_valid_one_char(xs[0..0])
    return b
  end
  return count_interpretations(xs[1..-1],
      is_valid_one_char(xs[0..0]) ? a + b : b,
      is_valid_two_char(xs[0..1]) ? a : 0)
end

loop do
  line = gets.strip
  break if line == "0"
  puts count_interpretations(line, 1, 0)
end
